Make Stackdriver Logging API endpoint configurable.

qingling128 opened the pull request.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -321,6 +323,9 @@ module InternalConstants
                  :default => nil,
                  :secret => true
 
+    # The URL of Stackdriver Logging API.
```
This theoretically appears in some generated parameter description somewhere. Add something like "`Right now this only works with the gRPC path (use_grpc = true).`"?

qingling128 replied:
Added.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -369,6 +374,12 @@ def configure(conf)
                   ' partial_success flag will be ignored.'
       end
 
+      unless @logging_api_url == DEFAULT_LOGGING_API_URL || @use_grpc
```
Add a TODO, along the lines of:
```
      # TODO(qingling128): Remove this warning after the support is added. Also
      # remove the comment in the description of this configuration.
```

qingling128 replied:
Added.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -369,6 +374,12 @@ def configure(conf)
+      unless @logging_api_url == DEFAULT_LOGGING_API_URL || @use_grpc
+        @log.warn 'Detected customized logging_api_url while use_grpc is not' \
+                  ' enabled. Customized logging_api_url for the non-gRPC path' \
+                  ' is not supported. The config will be ignored.'
```
`The logging_api_url option will be ignored.`.

qingling128 replied:
Added.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1718,7 +1729,7 @@ def api_client
         creds = ssl_creds.compose(creds)
         @client = Google::Logging::V2::LoggingServiceV2::Stub.new(
-          'logging.googleapis.com', creds)
+          @logging_api_url, creds)
```
Technically, in the case of an RPC stub, it's a _host_ (a standalone host name), not a _URL_ (`protocol://host:port/path`). In fact, we would need to reconcile the need to specify the host on the gRPC path and the full URL on the non-gRPC path.

qingling128 replied:
Maybe we can let them always specify the full URL like `https://logging.googleapis.com`. Then for the gRPC path, we throw away the `protocal` part?

WDYT?

igorpeshansky replied:
That could work. Will gRPC take a `host:port` spec? Also, what happens if the URL has a path? We should probably specify what the expected format is.

qingling128 commented:
Friendly ping.

qingling128 commented:
Rebased and added a protocal section to the url. Also added some example of what the url should be. Yes, `gRPC` takes `port` and `path` just fine.

I also added a flag `use_secure_channel` to skip authentication for Logging API mocks.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -323,6 +325,18 @@ module InternalConstants
+    # Whether to use a secure channel when talking to the Logging API. This
+    # should only be false if the Stackdriver Logging Agent is set up to talk
+    # to a mocked / stubbed Logging API. Otherwise the requests will fail.
+    config_param :use_secure_channel, :bool, :default => true
```
Isn't this equivalent to having the `https://` protocol on the `logging_api_url` (as opposed to the `http://` protocol)?

igorpeshansky replied:
What I mean is, should we use the `http://` vs. `https://` prefix to infer this value, rather than adding a separate config option?

igorpeshansky commented on test/plugin/test_out_google_cloud_grpc.rb:
```diff
@@ -226,21 +226,29 @@ def test_non_integer_timestamp
+        creds = GRPC::Core::CallCredentials.new(authentication.updater_proc)
+        ssl_creds.compose(creds)
+      else
+        :this_channel_is_insecure
```
Shouldn't you be assigning this to `creds`?

igorpeshansky replied:
I'm confused. The value by itself is useless, isn't it? What is this method even doing? Isn't it a no-op?

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -323,6 +325,18 @@ module InternalConstants
 
+    # The URL of Stackdriver Logging API. Right now this only works with the
+    # gRPC path (use_grpc = true).
+    # The default value is: https://logging.googleapis.com.
```
This sentence is redundant, as it's obvious from the `:default` item below.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -323,6 +325,18 @@ module InternalConstants
+    # gRPC path (use_grpc = true).
+    # The default value is: https://logging.googleapis.com.
+    # One common usage of this config is to provide a mocked / stubbed Logging
+    # API. e.g. http://localhost:52000.
```
Nit: comma after `API` and after `e.g.`.

igorpeshansky replied:
Careful -- your current code won't work with the `http://` prefix (it should, though).
Also, just wondering: do we really want to encode a port here?

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1768,13 +1791,17 @@ def init_api_client
         @client = Google::Cloud::Logging::V2::LoggingServiceV2Client.new(
           channel: GRPC::Core::Channel.new(
-            'logging.googleapis.com', nil, creds))
+            @logging_api_url.sub(%r{^https?\://}, ''), nil, creds))
```
Would it make sense to use `URI.parse` instead of string substitution?

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1768,13 +1791,17 @@ def init_api_client
+        else
+          creds = :this_channel_is_insecure
+        end
         @client = Google::Cloud::Logging::V2::LoggingServiceV2Client.new(
```
Does this need to be rebased again?

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1866,6 +1891,18 @@ def api_client
+      host = uri.host
+      unless host
+        raise Fluent::ConfigError,
+              'The URL specified by config logging_api_url is invalid:' \
```
`The logging_api_url option specifies an invalid URL: #{@logging_api_url}`.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1866,6 +1891,18 @@ def api_client
+      unless host
+        raise Fluent::ConfigError,
+              'The URL specified by config logging_api_url is invalid:' \
+              " #{logging_api_url}."
```
Are you missing an `@` here?

igorpeshansky commented on test/plugin/test_out_google_cloud_grpc.rb:
```diff
@@ -228,29 +228,28 @@ def create_driver(conf = APPLICATION_DEFAULT_CONFIG,
                     tag = 'test',
                     multi_tags = false)
     conf += USE_GRPC_CONFIG
+    use_secure_channel = !conf.include?('use_secure_channel false')
```
Isn't this stale now?

qingling128 replied:
Right. Removed this and all other occurrences.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1866,6 +1886,22 @@ def api_client
       @client
     end
 
+    def use_secure_channel
```
I think having this as a separate method detracts from readability. Would rather see this test inline.

qingling128 replied:
Inlined.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1866,6 +1886,22 @@ def api_client
     end
 
+    def use_secure_channel
+      @logging_api_url.start_with?('https')
```
Shouldn't this be `starts_with?('https://')`?

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1866,6 +1886,22 @@ def api_client
+      @logging_api_url.start_with?('https')
+    end
+
+    def grpc_channel_name
```
Probably makes sense to inline this method into `init_api_client` as well, allowing you to use `uri.scheme` instead of the `starts_with` string test.

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -325,6 +327,12 @@ module InternalConstants
+    # The URL of Stackdriver Logging API. Right now this only works with the
+    # gRPC path (use_grpc = true).
+    # One common usage of this config is to provide a mocked / stubbed Logging
+    # API, e.g., http://localhost:52000.
```
Probably worth mentioning that using `http://` will create an insecure channel.

qingling128 replied:
Added.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -325,6 +327,12 @@ module InternalConstants
                  :secret => true
 
+    # The URL of Stackdriver Logging API. Right now this only works with the
+    # gRPC path (use_grpc = true). An unsecured channel is used if the url
```
Nit: `URL`.

qingling128 replied:
Done.

qingling128 commented:
PTAL
